### PR TITLE
Feature/worker 2

### PR DESCRIPTION
**Pull request by andre-merzky**
This improves the task overlay:
  * better performance
  * custom comm channels
  * process level isolation
  * additional execution modes
  * node level scheduling

### CODE REVIEW

**Comment by andre-merzky** [comment]
This is getting closer to releasability - the code has been used in COVID production at scale, and while some scaling properties remain poorly understood, the overall behavior is stable.

**Comment by andre-merzky** [comment]
@iparask : Ioannis, thanks for looking through this!  Yes, I know it's again a somewhat involved PR.  Maybe it would be worthwhile to do another code walkthrough, together with @mtitov and @mturilli ?  That code will, in this form or another, stay with us for a while...

File: src/radical/pilot/task_overlay/master.py [hunk 1]
```diff
@@ -169,7 +247,103 @@ def wait(self, count=None, uids=None):
...
+            completed = [s for s in states if s in ['DONE', 'FAILED']]
+
+          # self._log.debug('%d =?= %d', len(completed), len(states))
+            if len(completed) == len(states):
+                break
+
+            time.sleep(5.0)
+        t_stop = time.time()
+
+        self._log.debug('master runtime: %.2fs', t_stop - t_start)
```
[the remark this answers is not in the document]
Comment (andre-merzky) [reply src/radical/pilot/task_overlay/master.py:292]: No, it's not.  That log should be removed actually - the master runs as a compute unit, and as such is timed with our normal profiler anyway.   
The `sleep(5)` is rather arbitrary of course.  I could not yet think of a decent way to have fast response times and avoid busy polling: one drawback of injecting tasks into the pipeline without the unit manager is that we cannot rely on the unit manager to issue state callbacks.  The only viable option I see at the moment is to listen to state events on the update pubsub and to issue callbacks for our workers ourself.  But that's for later.

File: src/radical/pilot/task_overlay/worker.py [hunk 2]
```diff
@@ -18,35 +24,524 @@ class Worker(rpu.Component):
...
+
+        # We need to make sure to run only up to `gpn` tasks using a gpu
+        # within that pool, so need a separate counter for that.
+        self._resources = {'cores' : [0] * self._n_cores,
+                           'gpus'  : [0] * self._n_gpus}
+
+        # resources are initially all free
+        self._res_evt.set()
+
+        pipes = len(list(glob.glob('/proc/%d/fd/*' % os.getpid())))
```
[the remark this answers is not in the document]
Comment (andre-merzky) [reply src/radical/pilot/task_overlay/worker.py:55]: Ah, thanks for pointing that out!  This needs to go, it's a debugging left-over where I tried to investigate some resource shortage on Frontera  (It turned out to be unrelated to number of open pipes at the end).

File: src/radical/pilot/task_overlay/worker.py [hunk 3]
```diff
@@ -18,35 +24,524 @@ class Worker(rpu.Component):
...
+
+            if gpus:
+                for n in range(self._n_gpus):
+                    if not self._resources['gpus'][n]:
+                        self._resources['gpus'][n] = 1
+                        alloc_gpus.append(n)
+                        if len(alloc_gpus) == gpus:
+                            break
+
+            assert(len(alloc_cores) == cores)
```
[the remark this answers is not in the document]
Comment (andre-merzky) [reply src/radical/pilot/task_overlay/worker.py:318]: Agree.

[the remark this answers is not in the document]
Comment (andre-merzky) [reply src/radical/pilot/task_overlay/worker.py:318]: Actually, it *does* make sense - the `for` loop above could run out of items before the core count is reached.

[the remark this answers is not in the document]
Comment (iparask) [reply src/radical/pilot/task_overlay/worker.py:318]: Still it doesn't. The resources are locked, there is a check that there are enough resources. As a result the for loop will never run out of items.

File: src/radical/pilot/task_overlay/worker.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (iparask) [reply src/radical/pilot/task_overlay/worker.py:55]: This exists in various places in the code.

[the remark this answers is not in the document]
Comment (andre-merzky) [reply src/radical/pilot/task_overlay/worker.py:55]: not anymore ;)

File: src/radical/pilot/task_overlay/worker.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (andre-merzky) [reply src/radical/pilot/task_overlay/worker.py:318]: Ack

File: src/radical/pilot/agent/executing/popen.py [hunk 4]
```diff
@@ -202,6 +202,7 @@ def spawn(self, launcher, cu):
             # Create string for environment variable setting
             env_string = ''
           # env_string += '. %s/env.orig\n'                % self._pwd
+            env_string += 'export RADICAL_BASE_DIR="%s"\n' % self._pwd
```
Comment (mtitov) [inline src/radical/pilot/agent/executing/popen.py:205]: @andre-merzky should we use `RADICAL_BASE` here? (according to [this](https://github.com/radical-cybertools/radical.utils/blob/devel/src/radical/utils/misc.py#L853))

Comment (andre-merzky) [reply src/radical/pilot/agent/executing/popen.py]: Ack, done.

**Comment by andre-merzky** [comment]
Sorry guys, I lost track of this one.  IIRC, we did the code walkthrough last, and so I think this PR should be ready for another review?

**Pull request merged**